doc: add flux-shell(1) manpage

grondo opened the pull request:
There's been a lot of questions lately about runtime configuration of the job shell, how to develop plugins, etc, so it seems like it is time to officially document some of these shell features.

This is a WIP because I just finished writing and haven't yet done an editing pass.

For now, all `flux-shell` features are documented in a single manpage, and some sections will need to be eventually fleshed out a bit more. However, for now this mostly has all the basic info. I also don't want to over-document at this time since some of these interfaces could change.

What is missing:

 * More detailed explanation of the shell's entries into the `guest.exec.eventlog`. (Is this needed?)
 * Illustrative examples of `initrc.lua`, and simple C and Lua shell plugins
 * This manpage references `flux_plugin_t` - we may need to also add a manpage for the `flux_plugin_*` functions
 * Anything else?

garlick commented:
Just had a quick read through this and looks great!  LMK when you're ready for a review pass.

grondo commented:
Ok, I've removed the WIP prefix now. I was going to add more examples to `flux-shell(1)`, but I got burned out for now.

While developing the manpage and adding small examples, I did run across a few other minor tweaks to the job shell that I added:
 
 * When developing an example I found it was too easy to mess up the arguments to the Lua `plugin.register()` function. The shell now throws an error if `plugin.register` is called, but no handlers are found.
 * It was inconvenient that flux-shell loads the `initrc.lua` even in "standalone" mode, since a lot of work the default initrc does doesn't make sense in that context. In fact, the first line of the default initrc was
    ```lua
    -- Skip initrc if running in standalone mode
    if shell.info.options.standalone then return end
    ```
    Now the flux-shell will *not* load an initrc in standalone mode, unless one is explicitly requested via the `--initrc` command line option or `initrc` jobspec shell option.
 * It is slightly inconvenient that use of the `initrc` jobspec option (`-o initrc=file.lua`) completely replaces the system default initrc, since that initrc may load system default plugins, etc.  Add a feature to the `initrc.lua` distributed with Flux that supports a *new* shell option `userrc`, which can be used to define a user-specified initrc that will run from the system default `initrc.lua`. The code is simply:
    ```lua
    if shell.options.userrc then source (shell.options.userrc) end
    ```

@SteVwonder, I requested your review since you are currently working on a shell plugin, so you may have feedback of what other docs we may need in the near term here.

SteVwonder commented on doc/test/spell.en.pws:
```diff
@@ -497,3 +497,18 @@ datetime
+initrc
+INITRC
+loglevel
+overwite
```
I think this is a spelling mistake: `overwite` -> `overwrite`.

grondo replied:
Ouch! Good catch!

grondo commented:
>  When $prefix is the prefix that flux-core was installed into, it works beautifully. I'm mainly concerned about setups like Spack, where the two repos are installed into separate directories. Should flux-sched provide it's own initrc.lua that sources it's installed plugins? I hesitate to suggest yet another environment variable, but maybe paths in an env var could be added to plugin.searchpath? Thoughts?

Good question! We should open an issue on this because I think it requires more thought, and I think there is still a lot of room for improvement in how the shell initializes. The "initrc" was designed to work a lot like a UNIX shell, where some system `initrc` is always sourced (e.g. /etc/profile), perhaps followed up by user-specific `initrc` (e.g. .bashrc, .profile, etc), but you are right this doesn't work well in the context of Spack. Are there any existing examples in Spack now that we could look at to inform our design here?

I'm not necessarily opposed to another environment variable. I always do try to use them as a last resort, though, because especially with Flux they will leak into subinstances, which may not be what you want... Maybe there is a better way to do it .

grondo commented:
Ok @SteVwonder, I pushed a fix for the typo in spelling dictionary (plus the typo it masked in flux-shell.rst). Thanks!

Setting MWP.

garlick commented on doc/man1/flux-shell.rst:
```diff
@@ -0,0 +1,446 @@
+the startup and execution of user jobs. flux-shell(1) runs as the job user,
+reads the jobspec and assigned resource set R for the job from the KVS,
+and using this data determines what local job tasks to execute. While job
+tasks are running, the job shell acts as the interface between the parent
+instance and the job by handling I/O, signals, and finally collecting
```
"Parent instance" might be slightly confusing here since we're not talking about nested instances (necessarily).  Would it be better to just say "Flux" or "Flux instance"?

grondo replied:
Good catch!

garlick commented on doc/man1/flux-shell.rst:
```diff
@@ -0,0 +1,446 @@
+
+The design of the Flux job shell allows customization through a set of
+builtin and runtime loadable shell plugins. These plugins are used to
+handle I/O redirection, PMI, CPU and GPU affinity, debugger support and
```
Would it be slightly clearer to say "standard I/O redirection"?

garlick commented on doc/man1/flux-shell.rst:
```diff
@@ -0,0 +1,446 @@
+in Lua. See the SHELL INITRC section below for details of the ``initrc``
+format and features.
+
+While the flux-shell(1) is a normal executable and may be run in
+standalone mode (``-s, --standalone``), this option and other command
+line options listed below are mainly used for testing purposes. Most
+Flux users will interact with flux-shell(1) through the execution of
+Flux jobs. Therefore, the OPTIONS section may be of less interest than
+the subsequent sections in this manual.
```
Maybe this could be shortened to:
> Most Flux users will interact with flux-shell(1) indirectly through the execution of Flux jobs, however flux-shell(1) accepts the following options in standalone mode (``-s, --standalone``), for test purposes.

grondo replied:
That is much better, thanks!

grondo commented:
Thanks for taking another look @garlick! I applied all your suggestions above (squashed).

I'll set MWP.

grondo commented:
Oh, I also added some text about `FLUX_KVS_NAMESPACE` in the `OPERATION` section, and a note in the `PLUGINS` section regarding access to Flux services from job shell plugins.
